src/mb_enhanced_cover_art_uploads/providers/archive.ts

```diff
@@ -0,0 +1,133 @@
+            }
+        }
+
+        return this.extractGenericImages(itemMetadata, baseDownloadUrl);
+    }
+
+    /**
+     * Entrypoint for MB/CAA providers to delegate to IA. Does not fall back
+     * onto generic extraction.
+     */
+    async findImagesCAA(itemId: string): Promise<CoverArt[]> {
+        const itemMetadata = await this.getItemMetadata(itemId);
+        const baseDownloadUrl = this.createBaseDownloadUrl(itemMetadata);
+        return this.extractCAAImages(itemId, baseDownloadUrl);
+    }
+
+    private async extractCAAImages(itemId: string, baseDownloadUrl: string): Promise<CoverArt[]> {
+        // Grabbing metadata through CAA isn't 100% reliable, since the info
+        // in the index.json isn't always up-to-date (see CAA-129, only a few
+        // cases though).
+        const caaIndexUrl = `https://archive.org/download/${itemId}/index.json`;
+        const caaIndexResp = await gmxhr(caaIndexUrl);
+        const caaIndex = safeParseJSON<CAAIndex>(caaIndexResp.responseText, 'Could not parse index.json');
+
+        return caaIndex.images.map((img) => {
+            const imageFileName = urlBasename(img.image);
+            return {
+                url: new URL(baseDownloadUrl + `${itemId}-${imageFileName}`),
+                comment: img.comment,
+                types: img.types.map((type) => ArtworkTypeIDs[type as keyof typeof ArtworkTypeIDs]),
+            };
+        });
+    }
+
+    extractGenericImages(itemMetadata: ArchiveMetadata, baseDownloadUrl: string): CoverArt[] {
+        const originalImagePaths = itemMetadata.files
+            .filter((file) => file.source === 'original' && ArchiveProvider.IMAGE_FILE_FORMATS.includes(file.format))
+            .map((file) => file.name);
+
+        return originalImagePaths.map((path) => {
+            return {
+                url: new URL(baseDownloadUrl + path),
+            };
+        });
+    }
+
+    private async getItemMetadata(itemId: string): Promise<ArchiveMetadata> {
+        const itemMetadataResp = await this.fetchPage(new URL(`https://archive.org/metadata/${itemId}`));
+        const itemMetadata = safeParseJSON<ArchiveMetadata>(itemMetadataResp, 'Could not parse IA metadata');
+
+        // IA's metadata API always returns a 200, even for items which don't
+        // exist.
+        if (!itemMetadata.server) {
+            throw new Error('Empty IA metadata, item might not exist');
+        }
+
+        if (itemMetadata.is_dark) {
+            throw new Error('Cannot extract images: This item is darked');
+        }
+
+        return itemMetadata;
+    }
+
+    private createBaseDownloadUrl(itemMetadata: ArchiveMetadata): string {
+        // While we could just use the standard archive.org/download/... URL,
+        // it would always lead to redirection warnings which can be avoided.
+        return `https://${itemMetadata.server}${itemMetadata.dir}/`;
```

Comment:
It's a shame that ES still has no obvious clean solution to join URL components without external libraries and you often have to resort to string manipulation.
```suggestion
    private createBaseDownloadUrl(itemMetadata: ArchiveMetadata): URL {
        // While we could just use the standard archive.org/download/... URL,
        // it would always lead to redirection warnings which can be avoided.
        return new URL(itemMetadata.dir, `https://${itemMetadata.server}`);
```
... would be a possible solution here (still not pretty because of the protocol), but as soon as you want to append path components to a non-root URL it starts to become ugly, e.g. `new URL('./' + path, baseDownloadUrl)` above. On the other hand ```new URL(`./${itemId}-${imageFileName}`, baseDownloadUrl)``` would look a lot better, so what do you think?

Reply:
`new URL(path, baseDownloadUrl)` should work fine, as long as `baseDownloadUrl` ends in a slash.

Reply:
I went with a small helper function because the URL constructor's argument order is backwards, which makes no sense IMO.

The helper could be further extended to add the slashes where necessary or even take a variable list of arguments to join an arbitrary number of URL parts, but that's not necessary at the moment and would require unit testing, and I'm feeling lazy and don't want write those today 😅